### PR TITLE
[Compute] New command: vm install-patches

### LINKED ISSUE #16790: New CLI command for install patches API
**Resource Provider**
Compute

**Description of Feature or Work Requested**
We need a new command for installPatches API, which is used to maintain security compliance on azure VMs

Command: az vm install-patches -g rg -n vm

Other input parameters as mentioned below:
NOTE: some parameters vary depending on the VM OS Type (i.e. Windows or Linux)

- maximumDuration (Required for both Windows and Linux VMs)
    Specifies the maximum amount of time that the operation will run. It must be an ISO 8601-compliant duration string such as PT4H (4 hours)

- rebootSetting (Required for both Windows and Linux VMs)
    Defines when it is acceptable to reboot a VM during a software update operation.
    Possible values are:
    
      - IfRequired
      - Never
      - Always

- classificationsToInclude (Optional for both Windows and Linux VMs however the possible values vary)
    The update classifications to select when installing patches. Possible values differ for Windows and Linux.

    Possible values for Windows VM:
      - Critical
      - Security
      - UpdateRollUp
      - FeaturePack
      - ServicePack
      - Definition
      - Tools
      - Updates
    
    Possible values for Linux VM:
      - Critical
      - Security
      - Other

- kbNumbersToInclude (Applicable to Windows VM only and is an **optional** parameter)
  KBs to include in the patch operation.

- kbNumbersToExclude (Applicable to Windows VM only and is an **optional** parameter)
  KBs to exclude in the patch operation.

- excludeKBsRequiringReboot (Applicable to Windows VM only and is an **optional** parameter)
  Filters out KBs that don't have a reboot behavior of 'NeverReboots' when this is set.

- packageNameMasksToInclude (Applicable to Linux VM only and is an **optional** parameter)
  Packages to include in the patch operation. Format: packageName_packageVersion.

- packageNameMasksToExclude (Applicable to Linux VM only and is an **optional** parameter)
  Packages to exclude in the patch operation. Format: packageName_packageVersion.

**Minimum API Version Required**
2020-12-01

**Swagger Link**
https://github.com/Azure/azure-rest-api-specs/blob/compute-2020-12/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L4038

**Target Date**
March 1st Public Preview (Ignite)

### PR DESCRIPTION
**Pull request by xfz11**
**Description**

New command vm install-patches
Resolve #16790 

***Example***
az vm install-patches -g MyResourceGroup -n MyVm --maximum-duration PT4H --reboot-setting IfRequired --classifications-to-include Critical Security --exclude-kbs-requiring-reboot true

**Testing Guide**

``` python
        self.cmd(
            'vm create -g {rg} -n vm --image Win2019Datacenter --enable-hotpatching true --admin-username azureuser --admin-password testPassword0 --nsg-rule NONE')
        self.cmd('vm install-patches -g {rg} -n vm --maximum-duration PT4H --reboot-setting IfRequired '
                 '--classifications-to-include Critical Security --exclude-kbs-requiring-reboot true', checks=[
            self.check('status', 'Succeeded')
        ])
```
**History Notes**

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by qwordy** [comment]
I suggest develop features in your own forked repository.

File: src/azure-cli/azure/cli/command_modules/vm/_actions.py [hunk 1]
```diff
@@ -237,3 +237,4 @@ def _get_latest_image_version(cli_ctx, location, publisher, offer, sku):
     if not top_one:
         raise CLIError("Can't resolve the version of '{}:{}:{}'".format(publisher, offer, sku))
     return top_one[0].name
+
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_actions.py:240]:
```suggestion
```

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 2]
```diff
@@ -385,6 +385,19 @@ def load_arguments(self, _):
     with self.argument_context('vm diagnostics set') as c:
         c.argument('storage_account', completer=get_resource_name_completion_list('Microsoft.Storage/storageAccounts'))
 
+    with self.argument_context('vm install-patches') as c:
+        c.argument('vm_name', arg_type=existing_vm_name)
+        c.argument('maximum_duration', type=str, help='Specifies the maximum amount of time that the operation will run. It must be an ISO 8601-compliant duration string such as PT4H (4 hours)')
+        c.argument('reboot_setting', type=str, help='Defines when it is acceptable to reboot a VM during a software update operation. Possible values are: IfRequired/ Never/Always')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:391]: Enum type is better.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 3]
```diff
@@ -385,6 +385,19 @@ def load_arguments(self, _):
     with self.argument_context('vm diagnostics set') as c:
         c.argument('storage_account', completer=get_resource_name_completion_list('Microsoft.Storage/storageAccounts'))
 
+    with self.argument_context('vm install-patches') as c:
+        c.argument('vm_name', arg_type=existing_vm_name)
+        c.argument('maximum_duration', type=str, help='Specifies the maximum amount of time that the operation will run. It must be an ISO 8601-compliant duration string such as PT4H (4 hours)')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:390]:
```suggestion
        c.argument('maximum_duration', type=str, help='Specify the maximum amount of time that the operation will run. It must be an ISO 8601-compliant duration string such as PT4H (4 hours)')
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 4]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+        raise CLIError('rebootSetting value should be IfRequired/Never/Always')
+    ccf = _compute_client_factory(cmd.cli_ctx)
+    vm = ccf.virtual_machines.get(resource_group_name, vm_name)
+    if not vm:
+        raise CLIError("Can't get the vm named {} in resource group {}".format(vm_name, resource_group_name))
+    osType = vm.storage_profile.os_disk.os_type
+    if osType == 'Windows':
+        if classifications_to_include:
+            for cti in classifications_to_include:
+                if cti not in VMGuestPatchClassificationWindows._value2member_map_:
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3748]: _value2member_map_ is a protected field

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 5]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+                    raise CLIError('classifications_to_include value for Windows VM should be Critical/Security/UpdateRollUp/FeaturePack/ServicePack/Definition/Tools/Updates')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3749]:
```suggestion
                    raise ValidationError('classifications_to_include value for Windows VM should be Critical/Security/UpdateRollUp/FeaturePack/ServicePack/Definition/Tools/Updates')
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 6]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+        windows_parameters = WindowsParameters(classifications_to_include=classifications_to_include,
+                                               kb_numbers_to_include=kb_numbers_to_include,
+                                               kb_numbers_to_exclude=kb_numbers_to_exclude,
+                                               exclude_kbs_requiring_reboot=exclude_kbs_requiring_reboot)
+        install_patches_input = VirtualMachineInstallPatchesParameters(maximum_duration=maximum_duration, reboot_setting=reboot_setting,
+                                                                       windows_parameters=windows_parameters)
+    elif osType == 'Linux':
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3756]:
```suggestion
    elif osType.lower() == 'linux':
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 7]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+    return re.match('^P(\d+Y)?(\d+M)?(\d+D)?(T(\d+H)?(\d+M)?(\d+S)?)?$', duration)
+
+# region install patches
+def install_vm_patches(cmd, resource_group_name, vm_name, maximum_duration, reboot_setting, classifications_to_include=None, kb_numbers_to_include=None, kb_numbers_to_exclude=None,
+                       exclude_kbs_requiring_reboot=None, package_name_masks_to_include=None, package_name_masks_to_exclude=None):
+    from azure.mgmt.compute.models import VirtualMachineInstallPatchesParameters, WindowsParameters, LinuxParameters, VMGuestPatchRebootSetting, VMGuestPatchClassificationWindows, VMGuestPatchClassificationLinux
+    if not ios8601Compaliant(maximum_duration):
+        raise CLIError('maximum_duration value should be an ISO 8601-compliant duration string')
+    if reboot_setting not in VMGuestPatchRebootSetting._value2member_map_:
+        raise CLIError('rebootSetting value should be IfRequired/Never/Always')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3739]: If rebootSetting is defined as a enum type, Azure CLI framework will handle this kind of error.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 8]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+    ccf = _compute_client_factory(cmd.cli_ctx)
+    vm = ccf.virtual_machines.get(resource_group_name, vm_name)
+    if not vm:
+        raise CLIError("Can't get the vm named {} in resource group {}".format(vm_name, resource_group_name))
+    osType = vm.storage_profile.os_disk.os_type
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3744]: You really wrote a lot of validations.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 9]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+        linux_parameters = LinuxParameters(classifications_to_include=classifications_to_include,
+                                           package_name_masks_to_include=package_name_masks_to_include,
+                                           package_name_masks_to_exclude=package_name_masks_to_exclude)
+        install_patches_input = VirtualMachineInstallPatchesParameters(maximum_duration=maximum_duration, reboot_setting=reboot_setting,
+                                                                       linux_parameters=linux_parameters)
+    else:
+        raise CLIError('osType {} of the vm is not allowed'.format(osType))
+
+
+    return ccf.virtual_machines.begin_install_patches(resource_group_name=resource_group_name, vm_name=vm_name, install_patches_input=install_patches_input)
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3770]: What's the average running time? --no-wait is what you need.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 10]
```diff
@@ -385,6 +385,19 @@ def load_arguments(self, _):
...
+                                                                 'ServicePack/Definition/Tools/Updates. '
+                                                                 'Possible values for Linux VM: Critical/Security/Security')
+        c.argument('kb_numbers_to_include', nargs='+', help='Space-separated list of KBs to include in the patch operation. Applicable to Windows VM only')
+        c.argument('kb_numbers_to_exclude', nargs='+', help='Space-separated list of KBs to exclude in the patch operation. Applicable to Windows VM only')
+        c.argument('exclude_kbs_requiring_reboot', arg_type=get_three_state_flag(), help="Filters out KBs that don't have a reboot behavior of 'NeverReboots' when this is set. Applicable to Windows VM only")
+        c.argument('package_name_masks_to_include', nargs='+', help='Space-separated list of packages to include in the patch operation. Format: packageName_packageVersion. Applicable to Linux VM only')
+        c.argument('package_name_masks_to_exclude', nargs='+', help='Space-separated list of packages to exclude in the patch operation. Format: packageName_packageVersion. Applicable to Linux VM only')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:399]: Any shorter names?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 11]
```diff
@@ -3724,3 +3724,49 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
 # endregion
+
+def ios8601Compaliant(duration):
+    import re
+    return re.match('^P(\d+Y)?(\d+M)?(\d+D)?(T(\d+H)?(\d+M)?(\d+S)?)?$', duration)
+
+# region install patches
+def install_vm_patches(cmd, resource_group_name, vm_name, maximum_duration, reboot_setting, classifications_to_include=None, kb_numbers_to_include=None, kb_numbers_to_exclude=None,
+                       exclude_kbs_requiring_reboot=None, package_name_masks_to_include=None, package_name_masks_to_exclude=None):
+    from azure.mgmt.compute.models import VirtualMachineInstallPatchesParameters, WindowsParameters, LinuxParameters, VMGuestPatchRebootSetting, VMGuestPatchClassificationWindows, VMGuestPatchClassificationLinux
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3735]: This usage is forbidden in Compute. Compute is a multi-API module. Please use a dynamic import method. For example,
```
    DiskEncryptionSet, EncryptionSetIdentity, KeyVaultAndKeyReference, SourceVault = cmd.get_models(
        'DiskEncryptionSet', 'EncryptionSetIdentity', 'KeyVaultAndKeyReference', 'SourceVault')
```

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 12]
```diff
@@ -5071,6 +5071,16 @@ def test_vm_ssh_key(self, resource_group):
         self.cmd('vm create -g {rg} -n vm3 --image centos --nsg-rule None --ssh-key-name k3 --generate-ssh-keys')
         self.cmd('sshkey show -g {rg} -n k3')
 
+class VMInstallPatchesScenarioTest(ScenarioTest):
+    @ResourceGroupPreparer(name_prefix='cli_test_vm_install_patches_')
+    def test_vm_install_patches(self, resource_group):
+        # Create new one
+        self.cmd(
+            'vm create -g {rg} -n vm --image Win2019Datacenter --enable-hotpatching true --admin-username azureuser --admin-password testPassword0 --nsg-rule NONE')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:5079]: Linux VM is not covered.

File: src/azure-cli/azure/cli/command_modules/vm/_help.py [hunk 13]
```diff
@@ -820,6 +820,15 @@
     text: az vm assess-patches -g MyResourceGroup -n MyVm
 """
 
+helps['vm install-patches'] = """
+type: command
+short-summary: Install patches on a VM.
+examples:
+  - name: Install patches on a VM.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_help.py:827]: [Optional] Add description of parameters and values.

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py [hunk 14]
```diff
@@ -146,7 +146,8 @@ def default_api_version(self):
             'galleries': '2019-12-01',
             'gallery_images': '2020-09-30',
             'gallery_image_versions': '2019-12-01',
-            'virtual_machine_scale_sets': '2020-12-01'
+            'virtual_machine_scale_sets': '2020-12-01',
+            'virtual_machine': '2020-12-01'
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/profiles/_shared.py:150]:
```suggestion
```

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 15]
```diff
@@ -385,6 +386,16 @@ def load_arguments(self, _):
     with self.argument_context('vm diagnostics set') as c:
         c.argument('storage_account', completer=get_resource_name_completion_list('Microsoft.Storage/storageAccounts'))
 
+    with self.argument_context('vm install-patches') as c:
+        c.argument('maximum_duration', type=str, help='Specify the maximum amount of time that the operation will run. It must be an ISO 8601-compliant duration string such as PT4H (4 hours)')
+        c.argument('reboot_setting', arg_type=get_enum_type(RebootSetting), help='Define when it is acceptable to reboot a VM during a software update operation. Possible values are: IfRequired/ Never/Always')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:391]:
```suggestion
        c.argument('reboot_setting', arg_type=get_enum_type(RebootSetting), help='Define when it is acceptable to reboot a VM during a software update operation.')
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 16]
```diff
@@ -3724,3 +3725,35 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
 
 # endregion
+
+
+# region install patches
+def install_vm_patches(cmd, resource_group_name, vm_name, maximum_duration, reboot_setting, classifications_to_include=None, kb_numbers_to_include=None, kb_numbers_to_exclude=None,
+                       exclude_kbs_requiring_reboot=None, package_name_masks_to_include=None, package_name_masks_to_exclude=None, no_wait=False):
+    VMInstallPatchesParameters, WindowsParameters, LinuxParameters, VMGuestPatchClassificationWindows, VMGuestPatchClassificationLinux = cmd.get_models(
+        'VirtualMachineInstallPatchesParameters', 'WindowsParameters', 'LinuxParameters', 'VMGuestPatchClassificationWindows', 'VMGuestPatchClassificationLinux')
+    ccf = _compute_client_factory(cmd.cli_ctx)
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3735]: client can be fetched from argument list.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 17]
```diff
@@ -3724,3 +3725,35 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+    vm = ccf.virtual_machines.get(resource_group_name, vm_name)
+    if not vm:
+        raise ValidationError("Can't get the vm named {} in resource group {}".format(vm_name, resource_group_name))
+    osType = vm.storage_profile.os_disk.os_type.lower()
+    if osType == 'windows':
+        if classifications_to_include:
+            for cti in classifications_to_include:
+                if cti not in [x.value for x in VMGuestPatchClassificationWindows]:
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3743]: Azure CLI framework can check enum validness for you.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 18]
```diff
@@ -3724,3 +3725,35 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+        install_patches_input = VMInstallPatchesParameters(maximum_duration=maximum_duration, reboot_setting=reboot_setting, windows_parameters=windows_parameters)
+    elif osType == 'linux':
+        if classifications_to_include:
+            for cti in classifications_to_include:
+                if cti not in [x.value for x in VMGuestPatchClassificationLinux]:
+                    raise ValidationError('classifications_to_include value for Windows VM should be Critical/Security/Other')
+        linux_parameters = LinuxParameters(classifications_to_include=classifications_to_include, package_name_masks_to_include=package_name_masks_to_include, package_name_masks_to_exclude=package_name_masks_to_exclude)
+        install_patches_input = VMInstallPatchesParameters(maximum_duration=maximum_duration, reboot_setting=reboot_setting, linux_parameters=linux_parameters)
+    else:
+        raise CLIError('osType {} of the vm is not allowed'.format(osType))
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3755]:
```suggestion
        raise ValidationError('osType {} of the vm is not allowed'.format(osType))
```

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 15 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py:391]: You added an extra space before `Never`

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 19]
```diff
@@ -3724,3 +3725,35 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+
+# region install patches
+def install_vm_patches(cmd, resource_group_name, vm_name, maximum_duration, reboot_setting, classifications_to_include=None, kb_numbers_to_include=None, kb_numbers_to_exclude=None,
+                       exclude_kbs_requiring_reboot=None, package_name_masks_to_include=None, package_name_masks_to_exclude=None, no_wait=False):
+    VMInstallPatchesParameters, WindowsParameters, LinuxParameters, VMGuestPatchClassificationWindows, VMGuestPatchClassificationLinux = cmd.get_models(
+        'VirtualMachineInstallPatchesParameters', 'WindowsParameters', 'LinuxParameters', 'VMGuestPatchClassificationWindows', 'VMGuestPatchClassificationLinux')
+    ccf = _compute_client_factory(cmd.cli_ctx)
+    vm = ccf.virtual_machines.get(resource_group_name, vm_name)
+    if not vm:
+        raise ValidationError("Can't get the vm named {} in resource group {}".format(vm_name, resource_group_name))
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3738]:
```suggestion
        raise ValidationError("Can't get the VM named {} in resource group {}".format(vm_name, resource_group_name))
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 20]
```diff
@@ -3724,3 +3725,35 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+    osType = vm.storage_profile.os_disk.os_type.lower()
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3739]: Will one of the object in path be None?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 21]
```diff
@@ -3724,3 +3725,36 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+
+
+# region install patches
+def install_vm_patches(cmd, client, resource_group_name, vm_name, maximum_duration, reboot_setting, classifications_to_include=None, kb_numbers_to_include=None, kb_numbers_to_exclude=None,
+                       exclude_kbs_requiring_reboot=None, package_name_masks_to_include=None, package_name_masks_to_exclude=None, no_wait=False):
+    VMInstallPatchesParameters, WindowsParameters, LinuxParameters, VMGuestPatchClassificationWindows, VMGuestPatchClassificationLinux = cmd.get_models(
+        'VirtualMachineInstallPatchesParameters', 'WindowsParameters', 'LinuxParameters', 'VMGuestPatchClassificationWindows', 'VMGuestPatchClassificationLinux')
+    vm = client.get(resource_group_name, vm_name)
+    if not vm:
+        raise ValidationError("Can't get the VM named {} in resource group {}".format(vm_name, resource_group_name))
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3737]: This line is not covered in test.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 22]
```diff
@@ -3724,3 +3725,36 @@ def set_disk_access(cmd, client, parameters, resource_group_name, disk_access_na
...
+        if classifications_to_include:
+            for cti in classifications_to_include:
+                if cti not in [x.value for x in VMGuestPatchClassificationWindows]:
+                    raise ValidationError('classifications_to_include value for Windows VM should be Critical/Security/UpdateRollUp/FeaturePack/ServicePack/Definition/Tools/Updates')
+        windows_parameters = WindowsParameters(classifications_to_include=classifications_to_include, kb_numbers_to_inclunde=kb_numbers_to_include, kb_numbers_to_exclude=kb_numbers_to_exclude, exclude_kbs_requirig_reboot=exclude_kbs_requiring_reboot)
+    elif osType == 'linux':
+        if classifications_to_include:
+            for cti in classifications_to_include:
+                if cti not in [x.value for x in VMGuestPatchClassificationLinux]:
+                    raise ValidationError('classifications_to_include value for Windows VM should be Critical/Security/Other')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3751]: Define a new type in cli core, list of enum, if you have time.

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 23]
```diff
@@ -5072,6 +5072,21 @@ def test_vm_ssh_key(self, resource_group):
         self.cmd('sshkey show -g {rg} -n k3')
 
 
+class VMInstallPatchesScenarioTest(ScenarioTest):
+    @ResourceGroupPreparer(name_prefix='cli_test_vm_install_patches_')
+    def test_vm_install_patches(self, resource_group):
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:5077]: Add intentional wrong commands to test your validation code.

File: src/azure-cli/azure/cli/command_modules/vm/commands.py [hunk 24]
```diff
@@ -260,7 +260,7 @@ def load_command_table(self, _):
         g.generic_update_command('update', custom_func_name='update_snapshot', setter_name='begin_create_or_update', setter_arg_name='snapshot', supports_no_wait=True)
         g.wait_command('wait')
 
-    with self.command_group('vm', compute_vm_sdk) as g:
+    with self.command_group('vm', compute_vm_sdk, client_factory=cf_vm) as g:
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/commands.py:263]: Dangerous. It affects all `vm` commands. Please make minimal change.

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py [hunk 25]
```diff
@@ -146,7 +146,7 @@ def default_api_version(self):
             'galleries': '2019-12-01',
             'gallery_images': '2020-09-30',
             'gallery_image_versions': '2019-12-01',
-            'virtual_machine_scale_sets': '2020-12-01'
+            'virtual_machine_scale_sets': '2020-12-01',
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/profiles/_shared.py:149]: Redundant `,`

**Review by qwordy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 10 above)
Comment (xfz11) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py:398]: origin long name has better readability as we discussed

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py (hunk 23 above)
Comment (xfz11) [reply src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:5077]: Validation codes are removed and enum validness will be checked by Azure CLI framework .

**Pull request merged**